File: sdks/go/baafoo/client_test.go
package baafoo

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

// 测试用 mock server
func newMockServer(t *testing.T) (*httptest.Server, *mockHandler) {
	mux := http.NewServeMux()
	h := &mockHandler{}

	mux.HandleFunc("/__baafoo__/api/agent/register", h.handleRegister)
	mux.HandleFunc("/__baafoo__/api/agent/heartbeat", h.handleHeartbeat)
	mux.HandleFunc("/__baafoo__/api/agent/poll", h.handlePoll)
	mux.HandleFunc("/__baafoo__/api/agent/recordings", h.handleRecordings)

	server := httptest.NewServer(mux)
	t.Cleanup(server.Close)
	return server, h
}

type mockHandler struct {
	registerCount  int32
	heartbeatCount int32
	pollCount      int32
	recordingCount int32
	lastRecordings []RecordingEntry
}

func (h *mockHandler) handleRegister(w http.ResponseWriter, r *http.Request) {
	h.registerCount++
	resp := map[string]interface{}{
		"success": true,
		"code":    200,
		"message": "OK",
		"data": map[string]interface{}{
			"agentId":         "test-agent-001",
			"mode":            "record-and-stub",
			"pollIntervalSec": 5,
		},
		"timestamp": time.Now().UnixMilli(),
	}
	writeJSON(w, resp)
}

func (h *mockHandler) handleHeartbeat(w http.ResponseWriter, r *http.Request) {
	h.heartbeatCount++
	resp := map[string]interface{}{
		"success":   true,
		"code":      200,
		"message":   "OK",
		"data":      nil,
		"timestamp": time.Now().UnixMilli(),
	}
	writeJSON(w, resp)
}

func (h *mockHandler) handlePoll(w http.ResponseWriter, r *http.Request) {
	h.pollCount++
	rules := []Rule{
		{
			ID:       "rule-001",
			Name:     "test-rule",
			Protocol: "http",
			Enabled:  true,
			Priority: 100,
			Conditions: []MatchCondition{
				{Field: "method", Operator: "equals", Value: "GET"},
				{Field: "path", Operator: "prefix", Value: "/api/"},
			},
			Responses: []ResponseEntry{
				{StatusCode: 200, Body: `{"ok":true}`},
			},
		},
	}
	resp := map[string]interface{}{
		"success": true,
		"code":    200,
		"message": "OK",
		"data": map[string]interface{}{
			"rules":   rules,
			"mode":    "stub",
			"version": time.Now().UnixMilli(),
		},
		"timestamp": time.Now().UnixMilli(),
	}
	writeJSON(w, resp)
}

func (h *mockHandler) handleRecordings(w http.ResponseWriter, r *http.Request) {
	var entries []RecordingEntry
	json.NewDecoder(r.Body).Decode(&entries)
	h.recordingCount += int32(len(entries))
	h.lastRecordings = entries
	resp := map[string]interface{}{
		"success":   true,
		"code":      200,
		"message":   fmt.Sprintf("Recorded %d", len(entries)),
		"data":      nil,
		"timestamp": time.Now().UnixMilli(),
	}
	writeJSON(w, resp)
}

func writeJSON(w http.ResponseWriter, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(v)
}

func TestNew(t *testing.T) {
	c := New(Options{ServerURL: "http://localhost:8084"})
	if c == nil {
		t.Fatal("expected non-nil client")
	}
	if c.GetMode() != ModeRecordAndStub {
		t.Errorf("expected default mode record-and-stub, got %s", c.GetMode())
	}
	if len(c.GetRules()) != 0 {
		t.Errorf("expected empty rules initially")
	}
}

func TestRegister(t *testing.T) {
	server, h := newMockServer(t)

	c := New(Options{
		ServerURL:            server.URL,
		Environment:          "test-env",
		HeartbeatIntervalSec: 1,
		PollIntervalSec:      1,
	})

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	if err := c.Start(ctx); err != nil {
		t.Fatalf("Start failed: %v", err)
	}
	defer c.Close()

	if c.GetAgentID() != "test-agent-001" {
		t.Errorf("expected agentId test-agent-001, got %s", c.GetAgentID())
	}
	// register 返回 record-and-stub，但初始 poll 可能更新为 stub
	mode := c.GetMode()
	if mode != ModeRecordAndStub && mode != ModeStub {
		t.Errorf("expected mode record-and-stub or stub, got %s", mode)
	}
	if h.registerCount != 1 {
		t.Errorf("expected 1 register call, got %d", h.registerCount)
	}
}

func TestPollRules(t *testing.T) {
	server, _ := newMockServer(t)

	c := New(Options{
		ServerURL:            server.URL,
		HeartbeatIntervalSec: 100,
		PollIntervalSec:      1,
	})

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	if err := c.Start(ctx); err != nil {
		t.Fatalf("Start failed: %v", err)
	}
	defer c.Close()

	// 等待 poll 执行
	time.Sleep(2 * time.Second)

	rules := c.GetRules()
	if len(rules) != 1 {
		t.Fatalf("expected 1 rule, got %d", len(rules))
	}
	if rules[0].ID != "rule-001" {
		t.Errorf("expected rule ID rule-001, got %s", rules[0].ID)
	}
	if c.GetMode() != ModeStub {
		t.Errorf("expected mode stub after poll, got %s", c.GetMode())
	}
}

func TestReportRecording(t *testing.T) {
	server, h := newMockServer(t)

	c := New(Options{
		ServerURL:            server.URL,
		HeartbeatIntervalSec: 100,
		PollIntervalSec:      100,
	})

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	if err := c.Start(ctx); err != nil {
		t.Fatalf("Start failed: %v", err)
	}
	defer c.Close()

	// 上报录制数据
	entry := RecordingEntry{
		Protocol:           "http",
		Host:               "example.com",
		Port:               8080,
		Method:             "GET",
		Path:               "/api/test",
		ResponseStatusCode: 200,
		ResponseBody:       `{"ok":true}`,
	}
	c.ReportRecording(entry)

	// 触发 flush
	c.Close()

	if h.recordingCount != 1 {
		t.Errorf("expected 1 recording, got %d", h.recordingCount)
	}
	if len(h.lastRecordings) != 1 {
		t.Fatalf("expected 1 recording in last batch, got %d", len(h.lastRecordings))
	}
	if h.lastRecordings[0].Method != "GET" {
		t.Errorf("expected method GET, got %s", h.lastRecordings[0].Method)
	}
}

func TestMatchRequest(t *testing.T) {
	c := New(Options{})
	c.rules.Store([]Rule{
		{
			ID:      "rule-001",
			Enabled: true,
			Conditions: []MatchCondition{
				{Field: "method", Operator: "equals", Value: "GET"},
				{Field: "path", Operator: "prefix", Value: "/api/"},
			},
		},
		{
			ID:      "rule-002",
			Enabled: true,
			Conditions: []MatchCondition{
				{Field: "method", Operator: "equals", Value: "POST"},
				{Field: "path", Operator: "prefix", Value: "/api/"},
			},
		},
	})

	rule := c.MatchRequest("GET", "/api/orders")
	if rule == nil {
		t.Fatal("expected match for GET /api/orders")
	}
	if rule.ID != "rule-001" {
		t.Errorf("expected rule-001, got %s", rule.ID)
	}

	rule = c.MatchRequest("POST", "/api/orders")
	if rule == nil {
		t.Fatal("expected match for POST /api/orders")
	}
	if rule.ID != "rule-002" {
		t.Errorf("expected rule-002, got %s", rule.ID)
	}

	rule = c.MatchRequest("DELETE", "/api/orders")
	if rule != nil {
		t.Error("expected no match for DELETE /api/orders")
	}
}

func TestAuthHeaders(t *testing.T) {
	server, _ := newMockServer(t)

	c := New(Options{
		ServerURL:            server.URL,
		APIKey:               "test-api-key",
		HeartbeatIntervalSec: 100,
		PollIntervalSec:      100,
	})

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	if err := c.Start(ctx); err != nil {
		t.Fatalf("Start failed: %v", err)
	}
	defer c.Close()

	// 如果 API Key 认证有问题，register 会返回错误
	// 这里仅验证 Start 不报错
}

File: sdks/nodejs/test/client.test.js
'use strict';

const http = require('http');
const assert = require('assert');
const { Client, Options } = require('../index');

// --- Mock Server ---

let registerCount = 0;
let heartbeatCount = 0;
let pollCount = 0;
let recordingCount = 0;
let lastRecordings = [];

function resetCounts() {
  registerCount = 0;
  heartbeatCount = 0;
  pollCount = 0;
  recordingCount = 0;
  lastRecordings = [];
}

const server = http.createServer((req, res) => {
  const url = new URL(req.url, `http://localhost`);
  const path = url.pathname;

  if (req.method === 'POST' && path === '/__baafoo__/api/agent/register') {
    registerCount++;
    sendJson(res, {
      success: true,
      code: 200,
      message: 'OK',
      data: {
        agentId: 'test-agent-001',
        mode: 'record-and-stub',
        pollIntervalSec: 5,
      },
      timestamp: Date.now(),
    });
  } else if (req.method === 'POST' && path === '/__baafoo__/api/agent/heartbeat') {
    heartbeatCount++;
    sendJson(res, {
      success: true,
      code: 200,
      message: 'OK',
      data: null,
      timestamp: Date.now(),
    });
  } else if (req.method === 'GET' && path === '/__baafoo__/api/agent/poll') {
    pollCount++;
    sendJson(res, {
      success: true,
      code: 200,
      message: 'OK',
      data: {
        rules: [
          {
            id: 'rule-001',
            name: 'test-rule',
            protocol: 'http',
            enabled: true,
            priority: 100,
            conditions: [
              { field: 'method', operator: 'equals', value: 'GET' },
              { field: 'path', operator: 'prefix', value: '/api/' },
            ],
            responses: [
              { statusCode: 200, body: '{"ok":true}' },
            ],
          },
        ],
        mode: 'stub',
        version: Date.now(),
      },
      timestamp: Date.now(),
    });
  } else if (req.method === 'POST' && path === '/__baafoo__/api/agent/recordings') {
    let body = '';
    req.on('data', (chunk) => { body += chunk; });
    req.on('end', () => {
      const entries = body ? JSON.parse(body) : [];
      recordingCount += entries.length;
      lastRecordings = entries;
      sendJson(res, {
        success: true,
        code: 200,
        message: `Recorded ${entries.length}`,
        data: null,
        timestamp: Date.now(),
      });
    });
  } else {
    sendJson(res, { success: false, code: 404, message: 'Not Found' }, 404);
  }
});

function sendJson(res, data, status = 200) {
  const body = JSON.stringify(data);
  res.writeHead(status, {
    'Content-Type': 'application/json',
    'Content-Length': Buffer.byteLength(body),
  });
  res.end(body);
}

// --- Tests ---

async function runTests() {
  // 启动 mock server
  await new Promise((resolve) => server.listen(0, resolve));
  const port = server.address().port;
  const baseUrl = `http://127.0.0.1:${port}`;

  function newClient(opts = {}) {
    return new Client(new Options({
      serverUrl: baseUrl,
      heartbeatIntervalSec: 100,
      pollIntervalSec: 100,
      ...opts,
    }));
  }

  let passed = 0;
  let failed = 0;

  async function test(name, fn) {
    resetCounts();
    try {
      await fn();
      console.log(`  PASS: ${name}`);
      passed++;
    } catch (e) {
      console.error(`  FAIL: ${name}`);
      console.error(`    ${e.message}`);
      failed++;
    }
  }

  console.log('Running Node.js SDK tests...\n');

  // Test 1: new client
  await test('new client has defaults', async () => {
    const c = newClient();
    assert.strictEqual(c.mode, 'record-and-stub');
    assert.strictEqual(c.rules.length, 0);
  });

  // Test 2: register
  await test('register gets agentId and mode', async () => {
    const c = newClient();
    await c.start();
    try {
      assert.strictEqual(c.agentId, 'test-agent-001');
      assert.ok(['record-and-stub', 'stub'].includes(c.mode), `unexpected mode: ${c.mode}`);
      assert.ok(registerCount >= 1, `expected registerCount >= 1, got ${registerCount}`);
    } finally {
      await c.close();
    }
  });

  // Test 3: poll rules
  await test('poll rules fetches and stores rules', async () => {
    const c = newClient({ pollIntervalSec: 1 });
    await c.start();
    try {
      await new Promise((resolve) => setTimeout(resolve, 2000));
      assert.strictEqual(c.rules.length, 1);
      assert.strictEqual(c.rules[0].id, 'rule-001');
      assert.strictEqual(c.mode, 'stub');
    } finally {
      await c.close();
    }
  });

  // Test 4: report recording
  await test('report recording uploads to server', async () => {
    const c = newClient();
    await c.start();
    try {
      c.reportRecording({
        protocol: 'http',
        host: 'example.com',
        port: 8080,
        method: 'GET',
        path: '/api/test',
        responseStatusCode: 200,
        responseBody: '{"ok":true}',
      });
    } finally {
      await c.close();
    }
    assert.strictEqual(recordingCount, 1, `expected 1 recording, got ${recordingCount}`);
    assert.strictEqual(lastRecordings.length, 1);
    assert.strictEqual(lastRecordings[0].method, 'GET');
  });

  // Test 5: match request
  await test('match request finds matching rule', async () => {
    const c = newClient();
    c._rules = [
      {
        id: 'rule-001',
        enabled: true,
        conditions: [
          { field: 'method', operator: 'equals', value: 'GET' },
          { field: 'path', operator: 'prefix', value: '/api/' },
        ],
      },
      {
        id: 'rule-002',
        enabled: true,
        conditions: [
          { field: 'method', operator: 'equals', value: 'POST' },
          { field: 'path', operator: 'prefix', value: '/api/' },
        ],
      },
    ];

    let rule = c.matchRequest('GET', '/api/orders');
    assert.ok(rule, 'expected match for GET /api/orders');
    assert.strictEqual(rule.id, 'rule-001');

    rule = c.matchRequest('POST', '/api/orders');
    assert.ok(rule, 'expected match for POST /api/orders');
    assert.strictEqual(rule.id, 'rule-002');

    rule = c.matchRequest('DELETE', '/api/orders');
    assert.strictEqual(rule, null, 'expected no match for DELETE /api/orders');
  });

  // Test 6: auth headers
  await test('auth headers do not break start', async () => {
    const c = newClient({ apiKey: 'test-api-key' });
    await c.start();
    try {
      assert.strictEqual(c.agentId, 'test-agent-001');
    } finally {
      await c.close();
    }
  });

  // 关闭 server
  server.close();

  console.log(`\n${passed} passed, ${failed} failed`);
  if (failed > 0) {
    process.exit(1);
  }
}

runTests().catch((e) => {
  console.error(e);
  process.exit(1);
});
